### ISSUE TITLE
Using unreserved keywords as identifiers results in unparsable section for Snowflake dialect

### ISSUE THREAD

**Issue by jonathanunderwood** [sqlfluff/sqlfluff#6991]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The following simple query, which uses `limit`, which is an _unreserved_ keyword, results in a failure of SQLFluff parsing:

```sql
select
    limit as renamed
from sometable;
```

gives: (note: templater disabled)
```bash
sqlfluff lint -t raw premium/models/intermediate/crowbar/junk.sql
== [premium/models/intermediate/crowbar/junk.sql] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'select'
L:   2 | P:   5 |  PRS | Line 2, Position 5: Found unparsable section: 'limit as
                       | renamed\nfrom sometable'
WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?
```

### Expected Behaviour

Since `limit` is an unreserved keyword I would expect this to be parsable. ~With SQLFluff 3.0.x it was - I noticed this problem on upgrading from 3.0.4 to 3.4.1, and it seems present in 3.1, 3.2, 3.3 as well as 3.4.1.~ [Edit: I was wrong about this]

### Observed Behaviour

see above.

### How to reproduce

See above.

### Dialect

Snowflake.

### Version

python 3.12.11
sqlfluff 3.4.1
sqlfluff-dbt-templater 3.4.1

### Configuration

Configured via `pyproject.toml`:

```toml
[tool.sqlfluff.core]
dialect = "snowflake"
templater = "dbt"
runaway_limit = 10
max_line_length = 135
indent_unit = "space"

[tool.sqlfluff.templater.dbt]
project_dir = "./premium"

[sqlfluff.templater.jinja]
load_macros_from_path = "premium/macros/"
apply_dbt_builtins = true

[tool.sqlfluff.indentation]
tab_space_size = 4

[tool.sqlfluff.layout.type.comma]
spacing_before = "touch:inline"
line_position = "trailing"

[tool.sqlfluff.rules.capitalisation.keywords]
extended_capitalisation_policy = "lower"

[tool.sqlfluff.rules.aliasing.table]
aliasing = "explicit"

[tool.sqlfluff.rules.aliasing.column]
aliasing = "explicit"

[tool.sqlfluff.rules.references.special_chars]
quoted_identifiers_policy = "aliases"
[tool.sqlfluff.rules.aliasing.expression]
allow_scalar = false

[tool.sqlfluff.rules.capitalisation.identifiers]
extended_capitalisation_policy = "lower"

[tool.sqlfluff.rules.capitalisation.functions]
extended_capitalisation_policy = "lower"

[tool.sqlfluff.rules.capitalisation.literals]
extended_capitalisation_policy = "lower"

[tool.sqlfluff.rules.references.keywords]
# Comma separated list of words to ignore for this rule
ignore_words = "type"

[tool.sqlfluff.rules.ambiguous.column_references]
group_by_and_order_by_style = "explicit"
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by jonathanunderwood** [comment]
Possibly related to #5420
